src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/DomainEventService.kt

```diff
@@ -207,35 +213,125 @@ class DomainEventService(
-        ),
-        personReference = SnsEventPersonReferenceCollection(
-          identifiers = listOf(
-            SnsEventPersonReference("CRN", crn),
-            SnsEventPersonReference("NOMS", nomsNumber),
-          ),
-        ),
+    if (!emitDomainEventsEnabled) {
+      log.info("Not emitting SNS event for domain event because domain-events.cas1.emit-enabled is not enabled")
+      return
+    }
+
+    if (asyncSaveEnabled) {
+      doAsyncSave(
+        domainEvent,
+        typeName,
+        typeDescription,
+        detailUrl,
+        crn,
+        nomsNumber,
       )
+    } else {
+      doSyncSave(
+        domainEvent,
+        typeName,
+        typeDescription,
+        detailUrl,
+        crn,
+        nomsNumber,
+      )
+    }
+  }
+
+  fun createSnsEvent(
+    domainEvent: DomainEvent<*>,
+    typeName: String,
+    typeDescription: String,
+    detailUrl: String,
+    crn: String,
+    nomsNumber: String,
+  ): SnsEvent = SnsEvent(
+    eventType = typeName,
+    version = 1,
+    description = typeDescription,
+    detailUrl = detailUrl,
+    occurredAt = domainEvent.occurredAt.atOffset(ZoneOffset.UTC),
+    additionalInformation = SnsEventAdditionalInformation(
+      applicationId = domainEvent.applicationId,
+    ),
+    personReference = SnsEventPersonReferenceCollection(
+      identifiers = listOf(
+        SnsEventPersonReference("CRN", crn),
+        SnsEventPersonReference("NOMS", nomsNumber),
+      ),
+    ),
+  )
 
-      val publishResult = domainTopic.snsClient.publish(
-        PublishRequest(domainTopic.arn, objectMapper.writeValueAsString(snsEvent))
-          .withMessageAttributes(mapOf("eventType" to MessageAttributeValue().withDataType("String").withStringValue(snsEvent.eventType))),
+  fun createPublishRequest(snsEvent: SnsEvent): PublishRequest =
+    PublishRequest(domainTopic.arn, objectMapper.writeValueAsString(snsEvent))
+      .withMessageAttributes(
+        mapOf(
+          "eventType" to MessageAttributeValue().withDataType("String").withStringValue(snsEvent.eventType),
+        ),
       )
 
-      log.info("Emitted SNS event (Message Id: ${publishResult.messageId}, Sequence Id: ${publishResult.sequenceNumber}) for Domain Event: ${domainEvent.id} of type: ${snsEvent.eventType}")
-    } else {
-      log.info("Not emitting SNS event for domain event because domain-events.cas1.emit-enabled is not enabled")
+  fun doAsyncSave(
+    domainEvent: DomainEvent<*>,
+    typeName: String,
+    typeDescription: String,
+    detailUrl: String,
+    crn: String,
+    nomsNumber: String,
+  ) = runBlocking {
+    val snsEvent = createSnsEvent(
+      domainEvent,
+      typeName,
+      typeDescription,
+      detailUrl,
+      crn,
+      nomsNumber,
+    )
+
+    val publishRequest = createPublishRequest(snsEvent)
+
+    retry(publishRequest, domainEvent, snsEvent)
+  }
+
+  suspend fun retry(publishRequest: PublishRequest, domainEvent: DomainEvent<*>, snsEvent: SnsEvent) {
```

Comment:
Might be better to give this a more descriptive name, something like `publishSnsEventWithRetry`